File: mobile/lib/domain/personalization/adaptive_question_selector.dart
import '../entities/question.dart';

class AnswerHistoryItem {
  final String questionId;
  final String skillId;
  final QuestionDifficulty difficulty;
  final bool isCorrect;

  const AnswerHistoryItem({
    required this.questionId,
    required this.skillId,
    required this.difficulty,
    required this.isCorrect,
  });
}

class AdaptiveQuestionSelector {
  /// Determines the target difficulty for the next question for a given skill.
  /// Standard progression:
  /// - Default / Initial: medium
  /// - If medium was answered correctly: hard
  /// - If medium or easy was missed: easy
  /// - If easy was answered correctly: medium
  /// - If hard was missed: medium
  static QuestionDifficulty determineNextDifficulty(
    String skillId,
    List<AnswerHistoryItem> history,
  ) {
    final skillHistory = history.where((h) => h.skillId == skillId).toList();
    if (skillHistory.isEmpty) {
      return QuestionDifficulty.medium;
    }

    final last = skillHistory.last;
    if (last.difficulty == QuestionDifficulty.medium) {
      return last.isCorrect ? QuestionDifficulty.hard : QuestionDifficulty.easy;
    } else if (last.difficulty == QuestionDifficulty.easy) {
      return last.isCorrect ? QuestionDifficulty.medium : QuestionDifficulty.easy;
    } else {
      // hard
      return last.isCorrect ? QuestionDifficulty.hard : QuestionDifficulty.medium;
    }
  }

  /// Selects the next adaptive question from the bank
  static Question? getNextQuestion(
    List<Question> questionBank,
    List<String> targetSkillIds,
    List<AnswerHistoryItem> history, [
    int targetQuestionsPerSkill = 2,
  ]) {
    final askedQuestionIds = history.map((h) => h.questionId).toSet();

    // 1. Balance question distribution across target skills
    final skillCounts = <String, int>{};
    for (final sid in targetSkillIds) {
      skillCounts[sid] = history.where((h) => h.skillId == sid).length;
    }

    // Find skills that haven't reached their quota
    final pendingSkills = targetSkillIds
        .where((sid) => (skillCounts[sid] ?? 0) < targetQuestionsPerSkill)
        .toList();

    if (pendingSkills.isEmpty) {
      return null;
    }

    // Prioritize the skill with fewest answered questions
    pendingSkills.sort((a, b) => (skillCounts[a] ?? 0).compareTo(skillCounts[b] ?? 0));
    final nextSkillId = pendingSkills.first;

    // 2. Determine target difficulty
    final targetDiff = determineNextDifficulty(nextSkillId, history);

    // 3. Find candidate questions in bank matching skill & difficulty
    var candidates = questionBank.where((q) {
      return q.skillId == nextSkillId &&
          q.difficulty == targetDiff &&
          !askedQuestionIds.contains(q.id);
    }).toList();

    // Fallback: any unanswered question in the same skill
    if (candidates.isEmpty) {
      candidates = questionBank.where((q) {
        return q.skillId == nextSkillId && !askedQuestionIds.contains(q.id);
      }).toList();
    }

    // Ultimate fallback: any unanswered question in bank
    if (candidates.isEmpty) {
      candidates = questionBank.where((q) => !askedQuestionIds.contains(q.id)).toList();
    }

    if (candidates.isEmpty) return null;
    return candidates.first;
  }
}

File: src/domain/personalization/AdaptiveQuestionSelector.ts
import { Question } from '../entities/Question';

export interface AnswerHistoryItem {
  questionId: string;
  skillId: string;
  difficulty: 'easy' | 'medium' | 'hard';
  isCorrect: boolean;
}

export class AdaptiveQuestionSelector {
  /**
   * Selects the next adaptive question based on balanced skill coverage and dynamic difficulty adjustment
   */
  public static getNextQuestion(
    questionBank: Question[],
    targetSkillIds: string[],
    history: AnswerHistoryItem[],
    targetQuestionsPerSkill: number = 2
  ): Question | null {
    const answeredIds = new Set(history.map((h) => h.questionId));
    const unasked = questionBank.filter((q) => !answeredIds.has(q.id));

    if (unasked.length === 0) return null;

    // 1. Find which skill currently has the fewest questions answered
    const skillCounts: Record<string, number> = {};
    targetSkillIds.forEach((sid) => {
      skillCounts[sid] = history.filter((h) => h.skillId === sid).length;
    });

    // Sort skills by least answered
    const sortedSkills = [...targetSkillIds].sort(
      (a, b) => (skillCounts[a] || 0) - (skillCounts[b] || 0)
    );

    const candidateSkillId = sortedSkills[0];
    if (skillCounts[candidateSkillId] >= targetQuestionsPerSkill) {
      // All skills reached target count
      return null;
    }

    // 2. Determine target difficulty for candidateSkillId based on recent performance
    const skillHistory = history.filter((h) => h.skillId === candidateSkillId);
    let targetDifficulty: 'easy' | 'medium' | 'hard' = 'medium';

    if (skillHistory.length > 0) {
      const recent = skillHistory.slice(-2);
      const correctCount = recent.filter((r) => r.isCorrect).length;

      if (correctCount === recent.length && recent.length >= 1) {
        // Performing well -> increase difficulty
        const lastDiff = recent[recent.length - 1].difficulty;
        if (lastDiff === 'easy') targetDifficulty = 'medium';
        else if (lastDiff === 'medium') targetDifficulty = 'hard';
        else targetDifficulty = 'hard';
      } else if (correctCount === 0) {
        // Struggling -> decrease difficulty
        const lastDiff = recent[recent.length - 1].difficulty;
        if (lastDiff === 'hard') targetDifficulty = 'medium';
        else if (lastDiff === 'medium') targetDifficulty = 'easy';
        else targetDifficulty = 'easy';
      } else {
        targetDifficulty = 'medium';
      }
    }

    // 3. Find matching question in unasked pool
    let match = unasked.find(
      (q) => q.skillId === candidateSkillId && q.difficulty === targetDifficulty
    );

    // Fallback if exact difficulty is exhausted
    if (!match) {
      match = unasked.find((q) => q.skillId === candidateSkillId);
    }

    // Fallback to any remaining unasked question
    if (!match && unasked.length > 0) {
      match = unasked[0];
    }

    return match || null;
  }
}
